src/Course/FileIO.hs

```diff
@@ -81,54 +81,52 @@ the contents of c
 
 -- Given the file name, and file contents, print them.
 -- Use @putStrLn@.
-printFile ::
-  FilePath
-  -> Chars
-  -> IO ()
-printFile =
-  error "todo: Course.FileIO#printFile"
+printFile :: FilePath -> Chars -> IO ()
+printFile path contents =
+    putStrLn $ "============ " ++ path ++ "\n" ++ contents
 
 -- Given a list of (file name and file contents), print each.
 -- Use @printFile@.
-printFiles ::
-  List (FilePath, Chars)
-  -> IO ()
-printFiles =
-  error "todo: Course.FileIO#printFiles"
+printFiles :: List (FilePath, Chars) -> IO ()
+printFiles files = execIoNoResult printTuple files
+  where
+    printTuple (path, chars) = printFile path chars
 
 -- Given a file name, return (file name and file contents).
 -- Use @readFile@.
-getFile ::
-  FilePath
-  -> IO (FilePath, Chars)
-getFile =
-  error "todo: Course.FileIO#getFile"
+getFile :: FilePath -> IO (FilePath, Chars)
+getFile path = ((,) path) <$> readFile path
 
 -- Given a list of file names, return list of (file name and file contents).
 -- Use @getFile@.
-getFiles ::
-  List FilePath
-  -> IO (List (FilePath, Chars))
-getFiles =
-  error "todo: Course.FileIO#getFiles"
+getFiles :: List FilePath -> IO (List (FilePath, Chars))
+getFiles paths = execIo getFile paths
```

Comment:
In actual haskell, IIRC `sequence` is using `Monad`, and `sequenceA` is what this course calls `sequence`, and both these functions have a version that ends in `_` that results in a `()`.

ie... 

`sequence :: (Monad m, Traversable t) => t (m a) -> m (t a)`
`sequenceA :: (Applicative f, Traversable t) => t (f a) -> f (t a)`
`sequence_ :: (Monad m, Foldable t) => t (m a) -> m ()`

Oh there doesn't appear to be a `sequenceA_`... and I don't recall there being a `void` function in the prelude... (I would just usually use `const () <$>`... but there's probably one in the `Control.Monad` module that I'm not aware of. I should investigate more of the base modules :)

Reply:
`A` is literally short for `Applicative`... coz of history... coz at the time `sequence` was already defined on `Monad` and we had no `Applicative`.

`_` is just a stupid shorthand. I think generally Haskellers are pretty lazy, like mathematicians... which is why we have `mapM` rather than `monadicMap` or `mapMonadically`, aside from the fact that you can't easily capture the full meaning of `mapM` in a couple of words easily... so it's just a hint that it's doing monadic things... coz customarily, the type variable used for monad is `m` ;-) It's all rather messy and imprecise, but that's the **disadvantage** of having an experimental language where we try out all kinds of different things... clean up takes sometimes many many years...

Reply:
Yeah. I understand. Once you generalise these patterns to a certain point, all your variable names become `a` and `b` because there's no name for them, really.

Reply:
Yeah... but you still want them to be distinct. So, `a` and `b` etc. are kind of annoying "names" because you have to say *something*.

But it'd be pretty tedious if we had to write `map : (some_input_value -> some_resultant_value) -> List some_input_value -> List some_output_value`, but also, sometimes a variable means something different in a different piece of the type signature... so it'd be utterly wrong for those cases.

Like say you wanted to understand this:

`lens :: (s -> a) -> (s -> b -> t) -> Lens s t a b`

So `lens` is a function that takes two functions and produces something called a `Lens` that can be used to focus on a certain structure. You can think of it as a piece of glass that can be inserted into other machines to focus on particular parts of a data structure.

we *could* rename it like this:

`lens :: (originalContext -> originalContent) -> (originalContext -> mappedContent -> mappedContext) -> Lens originalContext mappedContext originalContent mappedContent`

but it gets *super* confusing because of all the words... and it's not obvious that in this case "context" means "some wrappery/enclosing type that includes the content type".

And it doesn't help particularly that `Lens s t a b` is actually hiding what it actually is, so we can't get an intuition of what is "under the covers".

and if you actually *see* the definition of lens, it might not make it any clearer:

`type Lens s t a b = Functor f => (a -> f b) -> s -> f t`

and that's because... in order to understand this, one has to have a *really deep appreciation and understanding of what `Functor` really means*.

We can talk about lenses sometime because they're fascinating, but I wanted to use this as an illustration of the kind of problem of generic variable names on either side — when you do know, and when you don't.